### PR TITLE
Check for "illegal" IP nrs 0.0.0.0 and [::] before trying to make a c…

### PR DESCRIPTION
**Pull request by wvvelzen**
…onnection

As discussed, please consider this patch for inclusion in binkd.

Btw: I've only compiled and tested this on linux!

### CODE REVIEW

File: client.c [hunk 1]
```diff
@@ -299,6 +299,22 @@ static int call0 (FTN_NODE *node, BINKD_CONFIG *config)
...
+  static struct sockaddr_in sa_0;
+#ifdef AF_INET6
+  static struct sockaddr_in6 sa6_0;
+#endif
+
+  if (sa_init)
+  {
+    sa_init = 0;
+    memset((void *)&sa_0, 0, sizeof(struct sockaddr_in));
+    sa_0.sin_family = AF_INET;
```
Comment (pgul) [inline client.c]: This code is not thread-safe.
We may catch race condition if both threads check sa_init, then both set sa_init to 0, then:
1st: memset(), sa_0.sin_family = AF_INET
2nd: memset()
1st: running further with zero sin_family.

I think we have to init these structures before branch, may be in do_client() or clientmgr().

**Comment by pgul** [comment]
Thank you!
Please move sa_0 and sa6_0 initialization to another place (before branch client) and I'll merge it.

File: client.c (hunk 1 above)
Comment (wvvelzen) [inline client.c]: Hi Pavel,

On Mon, Jan 18, 2016 at 4:54 PM, Pavel notifications@github.com wrote:

> In client.c https://github.com/pgul/binkd/pull/4#discussion_r50013576:
> 
> > @@ -299,6 +299,22 @@ static int call0 (FTN_NODE *node, BINKD_CONFIG *config)
> >  #endif
> >    struct addrinfo *ai, *aiNodeHead, *aiHead, hints;
> >    int aiErr;
> > -  static int sa_init = 1;
> > -  static struct sockaddr_in sa_0;
> >   +#ifdef AF_INET6
> > -  static struct sockaddr_in6 sa6_0;
> >   +#endif
> >   +
> > -  if (sa_init)
> > -  {
> > -    sa_init = 0;
> > -    memset((void *)&sa_0, 0, sizeof(struct sockaddr_in));
> > -    sa_0.sin_family = AF_INET;
> 
> This code is not thread-safe.
> We may catch race condition if both threads check sa_init, then both set
> sa_init to 0, then:
> 1st: memset(), sa_0.sin_family = AF_INET
> 2nd: memset()
> 1st: running further with zero sin_family.
> 
> ​A, yes, of course, I didn't consider multitasking...

I think we have to init these structures before branch, may be in

> do_client() or clientmgr().
> 
> ​I'll look into it...

## 

Met vriendelijke groet / Best regards,
Wilfred van Velzen

**Comment by wvvelzen** [comment]
On Mon, Jan 18, 2016 at 5:05 PM, Pavel notifications@github.com wrote:

> Thank you!
> Please move sa_0 and sa6_0 initialization to another place (before branch
> client) and I'll merge it.
> 
> ​I might even make an array of these, so it would be easy to extend the
> list of invalid addresses if the need arises...

## 

Met vriendelijke groet / Best regards,
Wilfred van Velzen

**Pull request merged**